File: 388.longest-absolute-file-path.py
#
# @lc app=leetcode id=388 lang=python3
#
# [388] Longest Absolute File Path
#

# Time:  O(n)
# Space: O(d), d is the max depth of the paths

# @lc code=start
class Solution:
    def lengthLongestPath(self, input: str) -> int:
        max_len, path_len = 0, {0: 0}
        for line in input.split("\n"):
            name = line.lstrip("\t")
            depth = len(line) - len(name)
            if "." in name:  # file
                max_len = max(max_len, path_len[depth] + len(name))
            else:  # dir
                path_len[depth + 1] = path_len[depth] + len(name) + 1
        return max_len


# @lc code=end

File: 388.longest-absolute-file-path.js
/*
 * @lc app=leetcode id=388 lang=javascript
 *
 * [388] Longest Absolute File Path
 */

// @lc code=start
/**
 * @param {string} input
 * @return {number}
 */
var lengthLongestPath = function (input) {
  let maxLength = 0;
  const lookup = { 0: 0 };
  const lines = input.split("\n");
  for (const line of lines) {
    let depth = -1;
    while (line[++depth] === "\t");
    const name = line.slice(depth);
    if (name.includes(".")) {
      maxLength = Math.max(maxLength, lookup[depth] + name.length);
    } else {
      lookup[depth + 1] = lookup[depth] + name.length + 1;
    }
  }
  return maxLength;
};
// @lc code=end
